File: tobas_setup_assistant/src/tobas_setup_assistant/parameter_getters/double_spin_box.py
from overrides import override
from typing import Optional
from PyQt5.QtCore import Qt, pyqtSignal, pyqtSlot

from tobas_std_tools_py.string import convert_superscript
from tobas_rqt_tools.widgets import DoubleSpinBox

from .base import ParamGetterWidget


class ParamGetterWidget_DoubleSpinBox(ParamGetterWidget[float]):
    value_changed = pyqtSignal(float)

    def __init__(
        self,
        param_name: str,
        description_text: Optional[str] = None,
        decimals: int = 12,
        minimum: float = -1e9,
        maximum: float = +1e9,
        single_step: float = 1.0,
        default: Optional[float] = None,
        suffix: str = "",
    ) -> None:
        assert minimum < maximum
        assert single_step > 0.0
        assert decimals >= 0

        super().__init__(param_name, description_text)

        self._spin_box = DoubleSpinBox()
        self._rows.addWidget(self._spin_box)

        self._spin_box.setDecimals(decimals)  # 桁数の設定を最初にしないと，デフォルト値などが潰れてしまう
        self._spin_box.setMinimum(minimum)
        self._spin_box.setMaximum(maximum)
        self._spin_box.setSingleStep(single_step)
        if default is not None:
            assert minimum <= default <= maximum
            self._spin_box.setValue(default)
        self._spin_box.setSuffix(convert_superscript(suffix))
        self._spin_box.setFocusPolicy(Qt.StrongFocus)
        self._spin_box.valueChanged.connect(self._on_value_changed)

    @override
    def get(self) -> float:
        return self._spin_box.value()

    @override
    def set(self, src: float) -> None:
        self._spin_box.setValue(src)

    @pyqtSlot(float)
    def _on_value_changed(self, value: float) -> None:
        self.value_changed.emit(value)

File: tobas_gui/tobas_setup_assistant/src/param_getters/double_spin_box.cpp
#include <tobas_string_tools/core.hpp>

#include "tobas_setup_assistant/param_getters/double_spin_box.hpp"

namespace gui
{
namespace sa
{
ParamGetterWidget_DoubleSpinBox::ParamGetterWidget_DoubleSpinBox(
  const QString& param_name,
  const QString& description_text)
  : super(param_name, description_text)
{
  spin_box_ = new qt::DoubleSpinBox();
  rows_->addWidget(spin_box_);
  connect(spin_box_, QOverload<double>::of(&qt::DoubleSpinBox::valueChanged), this, &self::onValueChanged);
}

double ParamGetterWidget_DoubleSpinBox::getValue() const
{
  return spin_box_->value();
}

bool ParamGetterWidget_DoubleSpinBox::setValue(const double& src)
{
  if (src < spin_box_->minimum() || spin_box_->maximum() < src) {
    return false;
  }

  spin_box_->setValue(src);
  return true;
}

void ParamGetterWidget_DoubleSpinBox::setDecimals(int decimals)
{
  spin_box_->setDecimals(decimals);
}

void ParamGetterWidget_DoubleSpinBox::setMinimum(double minimum)
{
  spin_box_->setMinimum(minimum);
}

void ParamGetterWidget_DoubleSpinBox::setMaximum(double maximum)
{
  spin_box_->setMaximum(maximum);
}

void ParamGetterWidget_DoubleSpinBox::setSingleStep(double single_step)
{
  spin_box_->setSingleStep(single_step);
}

void ParamGetterWidget_DoubleSpinBox::setSuffix(const QString& suffix)
{
  spin_box_->setSuffix(QString::fromStdString(str::convertToSuperscript(suffix.toStdString())));
}

void ParamGetterWidget_DoubleSpinBox::onValueChanged(double value)
{
  Q_EMIT valueChanged(value);
}
}  // namespace sa
}  // namespace gui
